File: py/asymmetric/rsa.py
"""
RSA — Public-Key Cryptography
==============================
Full RSA from scratch: key generation, encrypt, decrypt, factorization attack.
Uses only Python's built-in integer arithmetic (no crypto libs).

Operations
----------
  generate  — generate RSA key pair for given bit size
  encrypt   — encrypt plaintext string with public key
  decrypt   — decrypt ciphertext integer with private key
  factorize — Pollard's rho factorization attack on n (for demo / small keys)
"""

import os
import math
import random
import secrets


# ─────────────────────────────────────────────
# Number Theory Helpers
# ─────────────────────────────────────────────

def _is_prime_miller_rabin(n: int, k: int = 20) -> bool:
    """Miller-Rabin primality test."""
    if n < 2: return False
    if n == 2 or n == 3: return True
    if n % 2 == 0: return False

    # Write n-1 as 2^r * d
    r, d = 0, n - 1
    while d % 2 == 0:
        r += 1
        d //= 2

    for _ in range(k):
        a = random.randrange(2, n - 1)
        x = pow(a, d, n)
        if x == 1 or x == n - 1:
            continue
        for _ in range(r - 1):
            x = pow(x, 2, n)
            if x == n - 1:
                break
        else:
            return False
    return True


def _generate_prime(bits: int) -> int:
    """Generate a random prime of exactly `bits` bits."""
    while True:
        n = secrets.randbits(bits) | (1 << (bits - 1)) | 1  # ensure odd, correct length
        if _is_prime_miller_rabin(n):
            return n


def _mod_inverse(e: int, phi: int) -> int:
    """Extended Euclidean Algorithm to find modular inverse."""
    g, x, _ = _extended_gcd(e, phi)
    if g != 1:
        raise ValueError("Modular inverse does not exist.")
    return x % phi


def _extended_gcd(a: int, b: int):
    if a == 0:
        return b, 0, 1
    g, x, y = _extended_gcd(b % a, a)
    return g, y - (b // a) * x, x


# ─────────────────────────────────────────────
# RSA Key Generation
# ─────────────────────────────────────────────

def _generate_keypair(bits: int):
    half = bits // 2
    p = _generate_prime(half)
    q = _generate_prime(half)
    while q == p:
        q = _generate_prime(half)

    n   = p * q
    phi = (p - 1) * (q - 1)

    # Common public exponent
    e = 65537
    if math.gcd(e, phi) != 1:
        e = 3
        while math.gcd(e, phi) != 1:
            e += 2

    d = _mod_inverse(e, phi)
    return {"p": p, "q": q, "n": n, "e": e, "d": d, "phi": phi}


# ─────────────────────────────────────────────
# OAEP-lite Padding (simplified, for demo)
# We use textbook RSA with a length prefix for educational clarity.
# ─────────────────────────────────────────────

def _text_to_int(text: str) -> int:
    return int(text.encode().hex(), 16)


def _int_to_text(n: int) -> str:
    h = hex(n)[2:]
    if len(h) % 2:
        h = "0" + h
    return bytes.fromhex(h).decode(errors="replace")


# ─────────────────────────────────────────────
# Pollard's Rho Factorization
# ─────────────────────────────────────────────

def _pollards_rho(n: int, max_iter: int = 100_000) -> int | None:
    """Returns a non-trivial factor of n, or None if not found."""
    if n % 2 == 0:
        return 2
    x = secrets.randbelow(n - 2) + 2
    y = x
    c = secrets.randbelow(n - 1) + 1
    d = 1
    steps = 0
    while d == 1 and steps < max_iter:
        x = (x * x + c) % n
        y = (y * y + c) % n
        y = (y * y + c) % n
        d = math.gcd(abs(x - y), n)
        steps += 1
    return d if d != n else None


def _factorize(n: int, timeout_iters: int = 20) -> tuple[int, int]:
    """Attempt to factor n into (p, q). Returns (0, 0) on failure."""
    for _ in range(timeout_iters):
        f = _pollards_rho(n)
        if f is None or not (1 < f < n):
            continue
        return f, n // f
    return 0, 0


# ─────────────────────────────────────────────
# Public Handle
# ─────────────────────────────────────────────

def handle(operation: str, params: dict) -> dict:
    op = operation.lower()

    if op == "generate":
        bits = int(params.get("key_size", 512))
        if bits not in (256, 512, 1024, 2048):
            bits = 512
        keys = _generate_keypair(bits)
        return {
            "public_key_n":  str(keys["n"]),
            "public_key_e":  str(keys["e"]),
            "private_key_d": str(keys["d"]),
            "prime_p":       str(keys["p"]),
            "prime_q":       str(keys["q"]),
            "phi_n":         str(keys["phi"]),
            "key_size":      f"{bits} bits",
        }

    elif op == "encrypt":
        plaintext = params.get("plaintext", "")
        n         = int(params.get("n", "0"))
        e         = int(params.get("e", "65537"))
        m         = _text_to_int(plaintext)
        if m >= n:
            raise ValueError("Message too large for key size.")
        c = pow(m, e, n)
        return {
            "ciphertext_int": str(c),
            "ciphertext_hex": hex(c),
        }

    elif op == "decrypt":
        c = int(params.get("ciphertext", "0"))
        n = int(params.get("n", "0"))
        d = int(params.get("d", "0"))
        m = pow(c, d, n)
        plaintext = _int_to_text(m)
        return {"plaintext": plaintext}

    elif op == "factorize":
        n = int(params.get("n", "0"))
        if _is_prime_miller_rabin(n):
            return {"result": f"{n} is prime — cannot factor."}
        p, q = _factorize(n)
        if p == 0:
            return {"result": "Factorization failed (n too large or took too long)."}
        if p * q != n:
            return {"result": "Factorization returned invalid factors."}
        # Recover private key
        e   = int(params.get("e", "65537"))
        phi = (p - 1) * (q - 1)
        try:
            d = _mod_inverse(e, phi)
            d_str = str(d)
        except Exception:
            d_str = "could not compute"
        return {
            "factor_p":          str(p),
            "factor_q":          str(q),
            "recovered_phi":     str(phi),
            "recovered_private": d_str,
        }

    else:
        raise ValueError(f"Unknown operation '{operation}' for RSA.")

File: internal/engine/asymmetric/rsa.go
// Package asymmetric implements RSA and ECC from scratch in Go.
package asymmetric

import (
	"crypto/rand"
	"fmt"
	"math/big"
	"strings"
)

// ─────────────────────────────────────────────
// RSA helpers
// ─────────────────────────────────────────────

var bigOne = big.NewInt(1)
var bigTwo = big.NewInt(2)

func generatePrime(bits int) (*big.Int, error) {
	return rand.Prime(rand.Reader, bits)
}

func textToInt(text string) *big.Int {
	return new(big.Int).SetBytes([]byte(text))
}

func intToText(n *big.Int) string {
	return string(n.Bytes())
}

// pollardRho attempts to find a non-trivial factor of n.
func pollardRho(n *big.Int) *big.Int {
	if new(big.Int).Mod(n, bigTwo).Sign() == 0 {
		return bigTwo
	}
	x, _ := rand.Int(rand.Reader, new(big.Int).Sub(n, bigTwo))
	x.Add(x, bigTwo)
	y := new(big.Int).Set(x)
	c, _ := rand.Int(rand.Reader, new(big.Int).Sub(n, bigOne))
	c.Add(c, bigOne)
	d := new(big.Int).Set(bigOne)

	f := func(v *big.Int) *big.Int {
		r := new(big.Int)
		r.Mul(v, v)
		r.Add(r, c)
		r.Mod(r, n)
		return r
	}

	maxIter := 100_000
	for d.Cmp(bigOne) == 0 && maxIter > 0 {
		maxIter--
		x = f(x)
		y = f(f(y))
		diff := new(big.Int).Sub(x, y)
		diff.Abs(diff)
		d.GCD(nil, nil, diff, n)
	}
	if maxIter == 0 || d.Cmp(n) == 0 {
		return nil
	}
	return d
}

// RSA is the public entry point for the RSA algorithm.
func RSA(operation string, params map[string]string) (map[string]string, error) {
	op := strings.ToLower(operation)

	switch op {
	case "generate":
		bits := 512
		switch params["key_size"] {
		case "256":
			bits = 256
		case "1024":
			bits = 1024
		case "2048":
			bits = 2048
		}
		half := bits / 2

		p, err := generatePrime(half)
		if err != nil {
			return nil, err
		}
		q, err := generatePrime(half)
		if err != nil {
			return nil, err
		}
		for q.Cmp(p) == 0 {
			q, err = generatePrime(half)
			if err != nil {
				return nil, err
			}
		}

		n := new(big.Int).Mul(p, q)
		phi := new(big.Int).Mul(
			new(big.Int).Sub(p, bigOne),
			new(big.Int).Sub(q, bigOne),
		)
		e := big.NewInt(65537)
		d := new(big.Int).ModInverse(e, phi)

		return map[string]string{
			"public_key_n":  n.String(),
			"public_key_e":  e.String(),
			"private_key_d": d.String(),
			"prime_p":       p.String(),
			"prime_q":       q.String(),
			"phi_n":         phi.String(),
			"key_size":      fmt.Sprintf("%d bits", bits),
		}, nil

	case "encrypt":
		n, ok1 := new(big.Int).SetString(params["n"], 10)
		e, ok2 := new(big.Int).SetString(params["e"], 10)
		if !ok1 || !ok2 {
			return nil, fmt.Errorf("invalid n or e")
		}
		m := textToInt(params["plaintext"])
		if m.Cmp(n) >= 0 {
			return nil, fmt.Errorf("message too large for key size")
		}
		c := new(big.Int).Exp(m, e, n)
		return map[string]string{
			"ciphertext_int": c.String(),
			"ciphertext_hex": fmt.Sprintf("0x%x", c),
		}, nil

	case "decrypt":
		c, ok1 := new(big.Int).SetString(params["ciphertext"], 10)
		n, ok2 := new(big.Int).SetString(params["n"], 10)
		d, ok3 := new(big.Int).SetString(params["d"], 10)
		if !ok1 || !ok2 || !ok3 {
			return nil, fmt.Errorf("invalid ciphertext, n, or d")
		}
		m := new(big.Int).Exp(c, d, n)
		return map[string]string{"plaintext": intToText(m)}, nil

	case "factorize":
		n, ok := new(big.Int).SetString(params["n"], 10)
		if !ok {
			return nil, fmt.Errorf("invalid n")
		}
		if n.ProbablyPrime(20) {
			return map[string]string{"result": fmt.Sprintf("%s is prime — cannot factor.", n)}, nil
		}
		var p *big.Int
		for i := 0; i < 20 && p == nil; i++ {
			p = pollardRho(n)
		}
		if p == nil {
			return map[string]string{"result": "Factorization failed (n too large or took too long)."}, nil
		}
		q := new(big.Int).Div(n, p)
		check := new(big.Int).Mul(p, q)
		if check.Cmp(n) != 0 {
			return map[string]string{"result": "Factorization returned invalid factors."}, nil
		}
		phi := new(big.Int).Mul(
			new(big.Int).Sub(p, bigOne),
			new(big.Int).Sub(q, bigOne),
		)
		e := big.NewInt(65537)
		d := new(big.Int).ModInverse(e, phi)
		dStr := "could not compute"
		if d != nil {
			dStr = d.String()
		}

		return map[string]string{
			"factor_p":          p.String(),
			"factor_q":          q.String(),
			"recovered_phi":     phi.String(),
			"recovered_private": dStr,
		}, nil

	default:
		return nil, fmt.Errorf("unknown operation %q for RSA", operation)
	}
}
